### PR TITLE
feat: Custom notifications and order modification price validation

### PR DESCRIPTION
**Pull request by tinaszheng**
This PR introduces a new way to fire one-off notifications and its first example usage is for the error notification when you're modifying an order from the TV chart with an invalid price.

### CODE REVIEW

File: src/hooks/useCustomNotification.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { addCustomNotification } from '@/state/notifications';
+
+export const useCustomNotification = () => {
+  const dispatch = useAppDispatch();
+
+  const notify = useCallback(
+    (customNotification: Omit<NotificationDisplayData, 'groupKey'>) => {
+      dispatch(
+        addCustomNotification({
+          id: uniqueId(),
```
Comment (tinaszheng) [inline src/hooks/useCustomNotification.ts:21]: open to other ID generators here too, anything random is fine

Comment (rosepuppy) [reply src/hooks/useCustomNotification.ts:21]: I have used `crypto?.randomUUID() ?? Date.now().toString()` elsewhere if we don't want to depend on lodash, but don't feel strongly

Comment (tinaszheng) [reply src/hooks/useCustomNotification.ts:21]: oh i like Date.now() :D

File: src/hooks/useCustomNotification.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  const dispatch = useAppDispatch();
+
+  const notify = useCallback(
+    (customNotification: Omit<NotificationDisplayData, 'groupKey'>) => {
+      dispatch(
+        addCustomNotification({
+          id: Date.now().toString(),
+          displayData: {
+            toastDuration: DEFAULT_TOAST_AUTO_CLOSE_MS,
+            groupKey: NotificationType.Custom,
```
Comment (rosepuppy) [inline src/hooks/useCustomNotification.ts:22]: Do we want all custom notifications to be grouped together? I think we can keep them separate, if we want to use it for various use cases. in which case we have `groupKey: id`

**Review by rosepuppy — APPROVED** [review]
Generally LGTM!

File: src/lib/orderModification.ts [hunk 3]
```diff
@@ -109,3 +109,15 @@ export const cancelOrderAsync = (
...
+
+export const isNewOrderPriceValid = (order: SubaccountOrder, newPrice: number) => {
+  const bookPrice = abacusStateManager.stateManager.state?.marketOrderbook(
+    order.marketId
+  )?.midPrice;
+  if (!bookPrice) return true;
+
+  const oldPrice = order.triggerPrice ?? order.price;
+
+  // Ensure newPrice makes the order remain on the same side of bookPrice
```
Comment (jaredvu) [inline src/lib/orderModification.ts:121]: > remain on the same side of bookPrice

nit: isn't `bookPrice` the `mid-market-price`? Update comment to `Ensure newPrice makes the order remain on the same side of the book`

File: src/lib/orderModification.ts [hunk 4]
```diff
@@ -109,3 +109,15 @@ export const cancelOrderAsync = (
...
+  return newPrice !== bookPrice && oldPrice - bookPrice > 0 === newPrice - bookPrice > 0;
```
Comment (jaredvu) [inline src/lib/orderModification.ts:122]: `oldPrice - bookPrice > 0 === newPrice - bookPrice > 0`

This shouldn't ever return `true` unless `oldPrice === newPrice` right?  Should the `===` -> `&&`?

File: src/hooks/useCustomNotification.ts (hunk 2 above)
Comment (jaredvu) [reply src/hooks/useCustomNotification.ts:22]: +1

File: src/lib/orderModification.ts [hunk 5]
```diff
@@ -109,3 +109,15 @@ export const cancelOrderAsync = (
     });
   });
 };
+
+export const isNewOrderPriceValid = (order: SubaccountOrder, newPrice: number) => {
```
Comment (jaredvu) [inline src/lib/orderModification.ts:113]: Can we brainstorm a good way to add this to unit tests `i.e. orderModification.spec.ts` considering we need abacus state?

File: src/lib/orderModification.ts (hunk 4 above)
Comment (tinaszheng) [reply src/lib/orderModification.ts:122]: its comparing the boolean values here - so basically if oldPrice < bookPrice, we also want newPrice < bookPrice, and vice versa

File: src/lib/orderModification.ts (hunk 3 above)
Comment (tinaszheng) [reply src/lib/orderModification.ts:121]: ah yes

File: src/hooks/useCustomNotification.ts (hunk 2 above)
Comment (tinaszheng) [reply src/hooks/useCustomNotification.ts:22]: yeah that makes sense!

File: src/lib/orderModification.ts (hunk 4 above)
Comment (jaredvu) [reply src/lib/orderModification.ts:122]: Ahhh i'm illiterate. tyty

**Comment by yogurtandjam** [comment]
generally LGTM - maybe instead of custom we can call them `UserTriggered` or something?

**Review by yogurtandjam — APPROVED** [review]

**Pull request merged**